metaflow/plugins/kubernetes/kubernetes.py

```diff
@@ -154,57 +153,285 @@ def launch_job(self, **kwargs):
+        jobset.environment_variable("MF_MASTER_PORT", str(port))
+        jobset.environment_variable("MF_WORLD_SIZE", str(num_parallel))
+        jobset.environment_variable_from_selector(
+            "JOBSET_RESTART_ATTEMPT",
+            "metadata.annotations['jobset.sigs.k8s.io/restart-attempt']",
+        )
+        jobset.environment_variable_from_selector(
+            "METAFLOW_KUBERNETES_JOBSET_NAME",
+            "metadata.annotations['jobset.sigs.k8s.io/jobset-name']",
+        )
+        jobset.environment_variable_from_selector(
+            "MF_WORKER_REPLICA_INDEX",
+            "metadata.annotations['jobset.sigs.k8s.io/job-index']",
+        )
+        ## -----Jobset specific env vars END here-----
+
+        tmpfs_enabled = use_tmpfs or (tmpfs_size and not use_tmpfs)
+        if tmpfs_enabled and tmpfs_tempdir:
+            jobset.environment_variable("METAFLOW_TEMPDIR", tmpfs_path)
+
+        for name, value in env.items():
+            jobset.environment_variable(name, value)
+
+        annotations = {
+            "metaflow/user": user,
+            "metaflow/flow_name": flow_name,
+            "metaflow/control-task-id": task_id,
+        }
+        if current.get("project_name"):
+            annotations.update(
+                {
+                    "metaflow/project_name": current.project_name,
+                    "metaflow/branch_name": current.branch_name,
+                    "metaflow/project_flow_name": current.project_flow_name,
+                }
+            )
 
-        _prefix = str(uuid4())[:6]
-        js = KubernetesClient().jobset(
-            name="js-%s" % _prefix,
+        for name, value in annotations.items():
+            jobset.annotation(name, value)
+
+        (
+            jobset.annotation("metaflow/run_id", run_id)
+            .annotation("metaflow/step_name", step_name)
+            .annotation("metaflow/attempt", attempt)
+            .label("app.kubernetes.io/name", "metaflow-task")
+            .label("app.kubernetes.io/part-of", "metaflow")
+        )
+
+        ## ----------- control/worker specific values START here -----------
+        # We will now set the appropriate command for the control/worker job
+        _get_command = lambda index, _tskid: self._command(
+            flow_name=flow_name,
             run_id=run_id,
-            task_id=task_id,
             step_name=step_name,
-            namespace=namespace,
-            labels=self._get_labels(labels),
-            annotations=annotations,
-            num_parallel=num_parallel,
-            job_spec=job_spec,
-            port=port,
+            task_id=_tskid,
+            attempt=attempt,
+            code_package_url=code_package_url,
+            step_cmds=[
+                step_cli.replace(
+                    METAFLOW_PARALLEL_STEP_CLI_OPTIONS_TEMPLATE,
+                    "--ubf-context $UBF_CONTEXT --split-index %s --task-id %s"
+                    % (index, _tskid),
```

Comment:
todo: try of `.format` instead of `.replace`

Reply:
@savingoyal : 

`.format` doesnt work as stated earlier for the following reason: 

The point in time we are running the `.format`, we already have constructed the step_cli with more format strings; For example at this point the step_cli looks like : 

```
python -u fail_flow.py --quiet --metadata service --environment local --datastore s3 --event-logger nullSidecarLogger --monitor nullSidecarMonitor --datastore-root s3://mybcket/metaflow/ --package-suffixes .py,.R,.RDS --pylint step train --run-id 5075 --input-paths ${METAFLOW_INPUT_PATHS_0} --retry-count 0 --max-user-code-retries 2 --namespace valay {METAFLOW_PARALLEL_STEP_CLI_OPTIONS_TEMPLATE}
```

At this point the step_cli already contains the `${METAFLOW_INPUT_PATHS_0}` which is a shell variable that will be replaced at runtime. At this point if we use `.format` then python will crash complaining that we haven't given any value for format string `METAFLOW_INPUT_PATHS_0`. So we use replace instead of format.